File: strategy/joint_optimizer.py
"""Joint pick optimizer: choose Entry A's and Entry B's picks together.

Unlike ``entry_a_value.py`` (optimizes Entry A alone) and
``entry_b_hedge.py`` (hedges Entry B against an already-fixed Entry A
pick), this module searches every valid ``(team_a, team_b)`` pair at once
and picks the one that maximizes:

    P(A wins) + P(B wins) - P(A loses AND B loses)

assuming independence between different games. The pair is always required
to come from two different games (see the constraints below), so that
independence assumption holds by construction -- we never need to reason
about two teams whose outcomes are perfectly correlated because they're
playing each other.

A pair is only considered if it:
  * never repeats a team either entry has already used this season
  * never picks the same team for both entries in the same week
  * never puts the two entries on opposing sides of the same game (since a
    game only has two teams, this is equivalent to: never pick two teams
    from the same game at all)
  * keeps Entry B's win probability at or above a configurable floor
    (default 65%), so B doesn't sacrifice safety purely to diversify away
    from A -- unless nothing clears the floor, in which case the floor is
    relaxed and that's called out in the reasoning rather than silently
    leaving Entry B without a pick.

The search space is small (at most ~2x the number of games per entry), so
this is a plain brute-force scan over all pairs, not an ILP/solver.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional

from data.models import Game
from models.win_prob import basis_phrase, resolve_team_win_probability
from state.entries_store import load_used_teams_for_entry
from strategy.entry_b_hedge import DEFAULT_MIN_WIN_PROB_FLOOR as DEFAULT_MIN_WIN_PROB_FLOOR_B
from strategy.entry_b_hedge import meets_win_prob_floor

ENTRY_A_NAME = "Entry A"
ENTRY_B_NAME = "Entry B"


@dataclass
class TeamOption:
    team_abbreviation: str
    opponent_abbreviation: Optional[str]
    event_id: Optional[str]
    win_pct: Optional[float]  # 0-100, may be spread-estimated
    win_pct_source: str  # "api" | "spread_estimate" | "unknown"
    spread_detail: Optional[str]


@dataclass
class PairScore:
    pick_a: TeamOption
    pick_b: TeamOption
    both_survive_pct: float
    one_survives_pct: float
    both_eliminated_pct: float
    objective_score: float


@dataclass
class JointSearchResult:
    best: Optional[PairScore]
    runner_up: Optional[PairScore]
    floor_relaxed: bool
    pairs_considered: int


@dataclass
class JointRecommendation:
    week: int
    pick_a: Optional[TeamOption]
    pick_b: Optional[TeamOption]
    both_survive_pct: Optional[float]
    one_survives_pct: Optional[float]
    both_eliminated_pct: Optional[float]
    reasoning: str
    floor_relaxed: bool = False
    pairs_considered: int = 0


def build_team_options(current_week_games: List[Game]) -> List[TeamOption]:
    """All teams playing a not-yet-started game this week, with win prob/spread."""
    options: List[TeamOption] = []
    for game in current_week_games:
        if game.state and game.state != "pre":
            continue
        spread_detail = game.odds.details if game.odds else None
        for team, opponent, is_home in ((game.home, game.away, True), (game.away, game.home, False)):
            if not team.abbreviation:
                continue
            resolved = resolve_team_win_probability(game, is_home)
            options.append(
                TeamOption(
                    team_abbreviation=team.abbreviation,
                    opponent_abbreviation=opponent.abbreviation,
                    event_id=game.event_id,
                    win_pct=resolved.win_pct,
                    win_pct_source=resolved.source,
                    spread_detail=spread_detail,
                )
            )
    return options


def _score_pair(a: TeamOption, b: TeamOption) -> PairScore:
    p_a = a.win_pct / 100.0
    p_b = b.win_pct / 100.0
    both_survive = p_a * p_b
    both_eliminated = (1 - p_a) * (1 - p_b)
    one_survives = 1.0 - both_survive - both_eliminated
    objective = p_a + p_b - both_eliminated
    return PairScore(
        pick_a=a,
        pick_b=b,
        both_survive_pct=both_survive * 100.0,
        one_survives_pct=one_survives * 100.0,
        both_eliminated_pct=both_eliminated * 100.0,
        objective_score=objective,
    )


def _same_game(a: TeamOption, b: TeamOption) -> bool:
    """Whether two options are opposite sides of one result.

    The event id answers this whenever there is one. Where there is not, the
    opponents do: two options are the two halves of one game exactly when each
    names the other as its opponent. `build_team_options` fills both fields
    together, so an option carrying one carries the other.

    The id alone used to be the whole test, which fails *open* -- a missing id
    on either side returned False, and the pair search read that as "different
    games". That is the guarantee this strategy is named for, and `_describe`
    goes on to print "Different games ..., so one result cannot end both"
    without re-checking, so the failure arrives with a sentence asserting the
    opposite. `_score_pair` mis-scores it too, computing (1-pA)(1-pB) for an
    outcome whose real probability is zero.

    ESPN has always sent `event.id`, so nobody has hit this. It is the
    degradation `safe_get` exists to survive: a renamed field becomes None
    rather than a crash, and the guards downstream are supposed to hold.
    """
    if a.event_id is not None and b.event_id is not None:
        return a.event_id == b.event_id
    return (
        (a.opponent_abbreviation is not None and a.opponent_abbreviation == b.team_abbreviation)
        or (b.opponent_abbreviation is not None and b.opponent_abbreviation == a.team_abbreviation)
    )


def find_best_pair(
    current_week_games: List[Game],
    used_teams_a: List[str],
    used_teams_b: List[str],
    min_win_prob_floor_b: float = DEFAULT_MIN_WIN_PROB_FLOOR_B,
) -> JointSearchResult:
    """Pure brute-force search over all constraint-satisfying (team_a, team_b) pairs."""
    options = build_team_options(current_week_games)

    available_a = [o for o in options if o.team_abbreviation not in used_teams_a and o.win_pct is not None]
    available_b_all = [o for o in options if o.team_abbreviation not in used_teams_b and o.win_pct is not None]
    available_b = [o for o in available_b_all if meets_win_prob_floor(o.win_pct, min_win_prob_floor_b)]

    floor_relaxed = False
    if not available_b and available_b_all:
        available_b = available_b_all
        floor_relaxed = True

    scored: List[PairScore] = []
    for a in available_a:
        for b in available_b:
            if a.team_abbreviation == b.team_abbreviation:
                continue
            if _same_game(a, b):
                continue  # same game -- opposing sides
            scored.append(_score_pair(a, b))

    scored.sort(key=lambda p: (-p.objective_score, p.pick_a.team_abbreviation, p.pick_b.team_abbreviation))

    best = scored[0] if scored else None
    # The same two teams with A/B swapped scores identically (the objective is
    # symmetric) but isn't a meaningfully different alternative -- skip past
    # any such swaps to find a genuinely different runner-up pairing, if one
    # exists.
    runner_up = None
    if best is not None:
        best_team_set = {best.pick_a.team_abbreviation, best.pick_b.team_abbreviation}
        for candidate in scored[1:]:
            if {candidate.pick_a.team_abbreviation, candidate.pick_b.team_abbreviation} != best_team_set:
                runner_up = candidate
                break
    return JointSearchResult(
        best=best, runner_up=runner_up, floor_relaxed=floor_relaxed, pairs_considered=len(scored)
    )


def _describe(option: TeamOption) -> str:
    win_pct = f"{option.win_pct:.1f}%" if option.win_pct is not None else "unknown"
    basis = basis_phrase(option.win_pct_source)
    spread = f", spread {option.spread_detail}" if option.spread_detail else ""
    return f"{option.team_abbreviation} vs {option.opponent_abbreviation or '?'} -- {win_pct} win prob{basis}{spread}"


def _build_reasoning(
    pair: PairScore, floor_relaxed: bool, min_win_prob_floor_b: float, runner_up: Optional[PairScore]
) -> str:
    """A few sentences, shown per pick on a phone and in the CLI report.

    Three things were dropped from this because the screen already carried
    them, and saying them twice is what made the panel a wall of text:

    * the both-survive / one-survives / both-eliminated split, which the view
      renders as its own factor rows above this prose -- see the `odds` array
      in the registry contract below.
    * "Entry B's pick clears the N% floor", which fired on every ordinary week
      and so told nobody anything. The interesting case is the floor being
      *relaxed*, and that still says so, loudly, and also raises a warning.
    * the objective scores behind the runner-up comparison. Which pairing won
      is worth knowing; that it scored 1.875 against 1.857 on a combined
      objective is not a number anybody can act on.

    Same treatment as the strategy notes in deadpool/src/engine/measured.js;
    see the comment above MEASURED there.
    """
    parts = [
        f"Entry A: {_describe(pair.pick_a)}.",
        f"Entry B: {_describe(pair.pick_b)}.",
    ]

    if floor_relaxed:
        parts.append(
            f"No team available to Entry B cleared the {min_win_prob_floor_b:.0f}% floor this week; "
            f"the floor was relaxed rather than leave Entry B without a pick."
        )

    parts.append(
        f"Different games (A faces {pair.pick_a.opponent_abbreviation or '?'}, "
        f"B faces {pair.pick_b.opponent_abbreviation or '?'}), so one result cannot end both."
    )

    if runner_up is not None:
        parts.append(
            f"Beat the next-best pairing, {runner_up.pick_a.team_abbreviation}/"
            f"{runner_up.pick_b.team_abbreviation}."
        )

    return " ".join(parts)


def recommend(
    current_week_games: List[Game],
    current_week: int,
    used_teams_a: Optional[List[str]] = None,
    used_teams_b: Optional[List[str]] = None,
    min_win_prob_floor_b: float = DEFAULT_MIN_WIN_PROB_FLOOR_B,
) -> JointRecommendation:
    """Jointly optimized picks for both entries, with reasoning.

    ``used_teams_a``/``used_teams_b`` default to loading
    ``state/used_teams_a.json`` / ``state/used_teams_b.json``.
    """
    if used_teams_a is None:
        used_teams_a = load_used_teams_for_entry(ENTRY_A_NAME)
    if used_teams_b is None:
        used_teams_b = load_used_teams_for_entry(ENTRY_B_NAME)

    search = find_best_pair(current_week_games, used_teams_a, used_teams_b, min_win_prob_floor_b)

    if search.best is None:
        return JointRecommendation(
            week=current_week,
            pick_a=None,
            pick_b=None,
            both_survive_pct=None,
            one_survives_pct=None,
            both_eliminated_pct=None,
            reasoning="No valid pick pair available this week (not enough eligible teams/games for both entries).",
            floor_relaxed=search.floor_relaxed,
            pairs_considered=search.pairs_considered,
        )

    best = search.best
    return JointRecommendation(
        week=current_week,
        pick_a=best.pick_a,
        pick_b=best.pick_b,
        both_survive_pct=best.both_survive_pct,
        one_survives_pct=best.one_survives_pct,
        both_eliminated_pct=best.both_eliminated_pct,
        reasoning=_build_reasoning(best, search.floor_relaxed, min_win_prob_floor_b, search.runner_up),
        floor_relaxed=search.floor_relaxed,
        pairs_considered=search.pairs_considered,
    )

File: deadpool/src/engine/strategies/joint-optimizer.js
/**
 * Both entries chosen together.
 *
 * Port of `strategy/joint_optimizer.py`, unchanged.
 *
 * Rather than fixing Entry A first and hedging B against it, this searches
 * every legal (teamA, teamB) pair and takes the one that maximises
 *
 *     P(A wins) + P(B wins) − P(A loses AND B loses)
 *
 * Independence between the two games is assumed, and — this is the part worth
 * keeping straight — it holds by construction rather than by hope: a pair is
 * only legal if the two teams come from different games, so there is never a
 * pair whose outcomes are perfectly anti-correlated because they are playing
 * each other.
 *
 * A pair is considered only if it
 *   * never repeats a team either entry has already used this season,
 *   * never gives both entries the same team,
 *   * never puts the two entries in the same game, and
 *   * keeps Entry B at or above its floor — unless nothing clears it, in which
 *     case the floor is relaxed and said out loud.
 *
 * The search space is at most about twice the number of games, squared, so
 * this is a plain scan over all pairs rather than a solver.
 */

import { buildOptions, sameGame, sameTeam, cmpStr, byWinPctDesc, boardBehind } from '../constraints.js';
import { f1, f0, f3 } from '../fmt.js';
import { basisPhrase } from '../win-prob.js';
import { DEFAULT_MIN_WIN_PROB_FLOOR, meetsWinProbFloor } from './entry-b-hedge.js';

export const ID = 'joint';
export { DEFAULT_MIN_WIN_PROB_FLOOR };

/** Every team playing a not-yet-started game, used teams included. */
export function buildTeamOptions(games) {
  return buildOptions(games).map((o) => ({
    teamAbbreviation: o.teamAbbreviation,
    teamName: o.teamName,
    opponentAbbreviation: o.opponentAbbreviation,
    eventId: o.eventId,
    startDate: o.startDate,
    winPct: o.winPct,
    winPctSource: o.winPctSource,
    winPctIsEstimated: o.winPctSource === 'spread_estimate',
    spreadDetail: o.spreadDetail,
  }));
}

export function scorePair(a, b) {
  const pA = a.winPct / 100.0;
  const pB = b.winPct / 100.0;
  const bothSurvive = pA * pB;
  const bothEliminated = (1 - pA) * (1 - pB);
  const oneSurvives = 1.0 - bothSurvive - bothEliminated;
  return {
    pickA: a,
    pickB: b,
    bothSurvivePct: bothSurvive * 100.0,
    oneSurvivesPct: oneSurvives * 100.0,
    bothEliminatedPct: bothEliminated * 100.0,
    objectiveScore: pA + pB - bothEliminated,
  };
}

export function findBestPair(games, usedTeamsA, usedTeamsB, minWinProbFloorB = DEFAULT_MIN_WIN_PROB_FLOOR) {
  const options = buildTeamOptions(games);

  const hasProb = (o) => o.winPct !== null && o.winPct !== undefined;
  const availableA = options.filter((o) => !usedTeamsA.includes(o.teamAbbreviation) && hasProb(o));
  const availableBAll = options.filter((o) => !usedTeamsB.includes(o.teamAbbreviation) && hasProb(o));
  let availableB = availableBAll.filter((o) => meetsWinProbFloor(o.winPct, minWinProbFloorB));

  let floorRelaxed = false;
  if (!availableB.length && availableBAll.length) {
    availableB = availableBAll;
    floorRelaxed = true;
  }

  const scored = [];
  for (const a of availableA) {
    for (const b of availableB) {
      if (sameTeam(a, b)) continue;
      if (sameGame(a, b)) continue;   // opposing sides of one result
      scored.push(scorePair(a, b));
    }
  }

  // (-objective, teamA, teamB) — the abbreviation tie-breaks are not cosmetic.
  // A board where several pairs share an objective is common once
  // probabilities come from spreads, and without them the answer would depend
  // on the order ESPN happened to list the games.
  scored.sort((x, y) =>
    (y.objectiveScore - x.objectiveScore)
    || cmpStr(x.pickA.teamAbbreviation, y.pickA.teamAbbreviation)
    || cmpStr(x.pickB.teamAbbreviation, y.pickB.teamAbbreviation));

  const best = scored.length ? scored[0] : null;

  // The same two teams with A and B swapped scores identically — the objective
  // is symmetric — but is not a meaningfully different alternative. Step past
  // any such swap to find a genuinely different runner-up pairing.
  let runnerUp = null;
  if (best) {
    const bestSet = new Set([best.pickA.teamAbbreviation, best.pickB.teamAbbreviation]);
    for (const candidate of scored.slice(1)) {
      const set = new Set([candidate.pickA.teamAbbreviation, candidate.pickB.teamAbbreviation]);
      if (set.size !== bestSet.size || [...set].some((t) => !bestSet.has(t))) { runnerUp = candidate; break; }
    }
  }

  return { best, runnerUp, floorRelaxed, pairsConsidered: scored.length };
}

export function describe(option) {
  const winPct = option.winPct === null || option.winPct === undefined ? 'unknown' : `${f1(option.winPct)}%`;
  const basis = basisPhrase(option.winPctSource);
  const spread = option.spreadDetail ? `, spread ${option.spreadDetail}` : '';
  return `${option.teamAbbreviation} vs ${option.opponentAbbreviation || '?'} -- ${winPct} win prob${basis}${spread}`;
}

/**
 * A few sentences, shown per pick on a phone and in the CLI report.
 *
 * Port of `build_reasoning` in strategy/joint_optimizer.py, held to it
 * character for character by the parity suite and fixtures/golden.
 *
 * Three things were dropped because the screen already carried them, and
 * saying them twice is what made the panel a wall of text:
 *
 *   * the both-survive / one-survives / both-eliminated split, which the view
 *     renders as its own factor rows above this prose — see `odds` below.
 *   * "Entry B's pick clears the N% floor", which fired on every ordinary week
 *     and so told nobody anything. The interesting case is the floor being
 *     *relaxed*, which still says so and also raises a warning.
 *   * the objective scores behind the runner-up comparison. Which pairing won
 *     is worth knowing; that it scored 1.875 against 1.857 is not actionable.
 *
 * Same treatment as the strategy notes in engine/measured.js.
 */
export function buildReasoning(pair, floorRelaxed, minWinProbFloorB, runnerUp) {
  const parts = [
    `Entry A: ${describe(pair.pickA)}.`,
    `Entry B: ${describe(pair.pickB)}.`,
  ];

  if (floorRelaxed) {
    parts.push(
      `No team available to Entry B cleared the ${f0(minWinProbFloorB)}% floor this week; `
      + 'the floor was relaxed rather than leave Entry B without a pick.',
    );
  }

  parts.push(
    `Different games (A faces ${pair.pickA.opponentAbbreviation || '?'}, `
    + `B faces ${pair.pickB.opponentAbbreviation || '?'}), so one result cannot end both.`,
  );

  if (runnerUp) {
    parts.push(
      `Beat the next-best pairing, ${runnerUp.pickA.teamAbbreviation}/`
      + `${runnerUp.pickB.teamAbbreviation}.`,
    );
  }

  return parts.join(' ');
}

export function recommend(games, currentWeek, usedTeamsA, usedTeamsB, minWinProbFloorB = DEFAULT_MIN_WIN_PROB_FLOOR) {
  const search = findBestPair(games, usedTeamsA, usedTeamsB, minWinProbFloorB);

  if (!search.best) {
    return {
      week: currentWeek,
      pickA: null, pickB: null,
      bothSurvivePct: null, oneSurvivesPct: null, bothEliminatedPct: null,
      reasoning: 'No valid pick pair available this week (not enough eligible teams/games for both entries).',
      floorRelaxed: search.floorRelaxed,
      pairsConsidered: search.pairsConsidered,
    };
  }

  const best = search.best;
  return {
    week: currentWeek,
    pickA: best.pickA,
    pickB: best.pickB,
    bothSurvivePct: best.bothSurvivePct,
    oneSurvivesPct: best.oneSurvivesPct,
    bothEliminatedPct: best.bothEliminatedPct,
    objectiveScore: best.objectiveScore,
    reasoning: buildReasoning(best, search.floorRelaxed, minWinProbFloorB, search.runnerUp),
    floorRelaxed: search.floorRelaxed,
    pairsConsidered: search.pairsConsidered,
  };
}

/* ------------------------------------------------ the registry contract -- */

export default {
  id: ID,
  name: 'Best pair, chosen together',
  // This said "the only one that can put your two entries on opposite sides of
  // the same game", which is the exact inverse of what the file it sits in
  // does: `sameGame(a, b)` is skipped eleven lines above, so this is the only
  // strategy here that *cannot* reach that holding. It was wrong on the picker
  // screen, at the moment somebody chooses — anyone who wanted that hedge
  // would have been sent to the one strategy structurally refusing to give it.
  //
  // What it actually guarantees is the opposite and is worth saying plainly:
  // the two picks always come from different games, so no single result can
  // take both entries. That is a real property, and it is why the independence
  // assumption in the scoring holds by construction rather than by hope.
  blurb: 'Weighs every legal pair of picks at once and takes the pair most likely to leave at '
    + 'least one entry standing. Always splits them across different games, so no single '
    + 'result can knock out both.',
  entries: 'both',
  params: [
    { key: 'minWinProbFloorB', label: 'Entry B floor', type: 'percent', default: DEFAULT_MIN_WIN_PROB_FLOOR, min: 0, max: 99, help: 'How safe the second entry has to be. Relaxed, loudly, if nothing clears it.' },
  ],

  run(ctx) {
    const [a, b] = ctx.entries;
    const floor = ctx.params.minWinProbFloorB ?? DEFAULT_MIN_WIN_PROB_FLOOR;
    const r = recommend(ctx.games, ctx.week, ctx.usedTeams[a.id] ?? [], ctx.usedTeams[b.id] ?? [], floor);

    const odds = r.pickA ? [
      { label: 'Both survive', value: `${f1(r.bothSurvivePct)}%`, weight: 1 },
      { label: 'One survives', value: `${f1(r.oneSurvivesPct)}%`, weight: 0 },
      { label: 'Both out', value: `${f1(r.bothEliminatedPct)}%`, weight: -1 },
    ] : [];

    const warnings = [];
    if (r.floorRelaxed) {
      warnings.push({ level: 'warn', text: `Nothing available to ${b.name} cleared the ${f0(floor)}% floor. The floor was relaxed rather than leave the entry without a pick.` });
    }
    if (!r.pickA) {
      warnings.push({ level: 'danger', text: 'No legal pair exists this week — not enough eligible teams in different games for both entries.' });
    }

    return {
      strategyId: ID,
      picks: [
        { entry: a.id, candidate: r.pickA, reasoning: r.reasoning, factors: r.pickA ? [
          { label: 'Win probability', value: `${f1(r.pickA.winPct)}%`, weight: 1, note: r.pickA.winPctIsEstimated ? 'Estimated from the spread.' : 'From ESPN.' },
          ...odds,
        ] : [] },
        { entry: b.id, candidate: r.pickB, reasoning: r.reasoning, factors: r.pickB ? [
          { label: 'Win probability', value: `${f1(r.pickB.winPct)}%`, weight: 1, note: r.pickB.winPctIsEstimated ? 'Estimated from the spread.' : 'From ESPN.' },
          ...odds,
        ] : [] },
      ],
      // A joint strategy has no per-entry ranking to offer — its unit of
      // choice is the pair, not the team — so the alternatives shown are the
      // week's board, ranked, for anyone overriding by hand.
      //
      // Pick-first, via the shared helper, and that is not cosmetic here. This
      // used to sort the board by win probability alone, which for a *pair*
      // search is not the same as putting the recommendation first: the best
      // pair routinely contains neither of the two best single teams. The view
      // drops index 0 as the recommendation, so it was dropping an arbitrary
      // team and offering the recommended one back as an alternative to
      // itself.
      candidates: {
        [a.id]: boardBehind(r.pickA, ctx.games, ctx.usedTeams[a.id] ?? []),
        [b.id]: boardBehind(r.pickB, ctx.games, ctx.usedTeams[b.id] ?? []),
      },
      considered: r.pairsConsidered,
      warnings,
      shared: { objectiveScore: r.objectiveScore ?? null, bothSurvivePct: r.bothSurvivePct, oneSurvivesPct: r.oneSurvivesPct, bothEliminatedPct: r.bothEliminatedPct },
    };
  },
};
